cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,194 @@
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import org.apache.twill.zookeeper.NodeData;
+import org.apache.twill.zookeeper.OperationFuture;
+import org.apache.twill.zookeeper.ZKClient;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * RouteStore where the routes are stored in a ZK persistent node. This is intended for use in distributed mode.
+ */
+public class ZKRouteStore implements RouteStore {
+  private static final Logger LOG = LoggerFactory.getLogger(ZKRouteStore.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_INTEGER_TYPE = new TypeToken<Map<String, Integer>>() { }.getType();
+  private static final int ZK_TIMEOUT_SECS = 5;
+
+  private final ZKClient zkClient;
+  private final ConcurrentMap<ProgramId, SettableFuture<RouteConfig>> routeConfigMap;
+
+  @Inject
+  public ZKRouteStore(ZKClient zkClient) {
+    this.zkClient = zkClient;
+    this.routeConfigMap = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public void store(final ProgramId serviceId, final RouteConfig routeConfig) {
+    Supplier<RouteConfig> supplier = Suppliers.ofInstance(routeConfig);
+    Future<RouteConfig> future = ZKExtOperations.createOrSet(zkClient, getZKPath(serviceId), supplier,
+                                                             ROUTE_CONFIG_CODEC, 10);
+    try {
+      future.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException ex) {
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public void delete(final ProgramId serviceId) throws NotFoundException {
+    OperationFuture<String> future = zkClient.delete(getZKPath(serviceId));
+    try {
+      future.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException ex) {
+      if (ex.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
```

Comment:
fetch shouldn't throw `NotFoundException`, since it is not required for the distribution settings to be there.